examples/run_group.py

```diff
@@ -0,0 +1,218 @@
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""This script allows a user to launch a chain of NFs from a JSON
+config file passed in as an argument when running the script"""
+
+import sys
+import json
+import os
+import shlex
+import time
+from signal import signal, SIGINT
+from subprocess import Popen
+from datetime import datetime
+
+def handler(signal_received, frame):
+    """Handles cleanup on user shutdown"""
+    # pylint: disable=unused-argument
+    for lf in log_files:
+        lf.close()
+    print("\nExiting...")
+    sys.exit(0)
+
+def get_config():
+    """Gets path to JSON config file"""
+    if len(sys.argv) < 2:
+        print("Error: Specify config file")
+        sys.exit(0)
+
+    file_name = sys.argv[1]
+    if os.path.exists(file_name):
+        return file_name
+    print("Error: No config file found")
+    sys.exit(0)
+
+def on_failure():
+    """Handles shutdown on error"""
+    for lf in log_files:
+        lf.close()
+    try:
+        script_pid = os.getpid()
+        pid_list = os.popen("ps -ef | awk '{if ($3 == " + str(script_pid) + ") print $2 " " $3}'")
+        pid_list = pid_list.read().split("\n")[:-2]
+        for pd in pid_list:
+            pd = pd.replace(str(script_pid), "")
+            temp = os.popen("ps -ef | awk '{if($3 == " + pd + ") print $2 " " $3}'")
+            temp = temp.read().replace(str(pd), "").replace("\n", "")
+            if temp != "":
+                _pid_for_nf = os.popen("ps -ef | awk '{if($3 == " + temp + ") print $2 " " $3}'")
+                _pid_for_nf = _pid_for_nf.read().replace(temp, "").replace("\n", "")
+                os.system("sudo kill " + _pid_for_nf)
+    except:
+        pass
+    print("Error occurred. Exiting...")
+    sys.exit(1)
+
+def on_timeout():
+    """Handles shutdown on timeout"""
+    for lf in log_files:
+        lf.close()
+    try:
+        script_pid = os.getpid()
+        pid_list = os.popen("ps -ef | awk '{if ($3 == " + str(script_pid) + ") print $2 " " $3}'")
+        pid_list = pid_list.read().split("\n")[:-2]
+        for pd in pid_list:
+            pd = pd.replace(str(script_pid), "")
+            temp = os.popen("ps -ef | awk '{if($3 == " + pd + ") print $2 " " $3}'")
+            temp = temp.read().replace(str(pd), "").replace("\n", "")
+            if temp != "":
+                _pid_for_nf = os.popen("ps -ef | awk '{if($3 == " + temp + ") print $2 " " $3}'")
+                _pid_for_nf = _pid_for_nf.read().replace(temp, "").replace("\n", "")
+                os.system("sudo kill " + _pid_for_nf)
+    except:
+        pass
+    print("Exiting on timeout...")
+    sys.exit(0)
+
+def running_services():
+    """Checks running NFs"""
+    if timeout != 0:
+        start = time.time()
+        time.clock()
+        elapsed = 0
+        while elapsed < timeout:
```

Comment:
This timeout variable is not checked for proper type. If someone makes a string, say `"10"`, instead of `10`, they'll get a nasty error message of a Python TypeError. Maybe there's a way to check this input? Seeing this with most inputs. We don't really have a check for the `directory` field either, or the parameters for NFs. It just takes the input and as long as it's reasonably correct JSON, the program tries to continue. However, it realizes quickly that the types provided do not match what's expected and crashes.